nav2_costmap_2d/src/collision_checker.cpp

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_costmap_2d/collision_checker.hpp"
+#include "nav2_costmap_2d/cost_values.hpp"
+#include "nav2_costmap_2d/exceptions.hpp"
+#include "nav2_costmap_2d/footprint.hpp"
+#include "nav2_util/line_iterator.hpp"
+
+using namespace std::chrono_literals;
+
+namespace nav2_costmap_2d
+{
+
+CollisionChecker::CollisionChecker(
+  std::shared_ptr<CostmapSubscriber> costmap_sub,
+  std::shared_ptr<FootprintSubscriber> footprint_sub,
+  std::string name)
+: name_(name),
+  costmap_sub_(costmap_sub),
+  footprint_sub_(footprint_sub)
+{}
+
+CollisionChecker::~CollisionChecker() {}
+
+bool CollisionChecker::isCollisionFree(
+  const geometry_msgs::msg::Pose2D & pose)
+{
+  try {
+    if (scorePose(pose) < 0) {
+      return false;
+    }
+    return true;
+  } catch (const IllegalPoseException & e) {
+    RCLCPP_ERROR(rclcpp::get_logger(name_), "%s", e.what());
+    return false;
+  } catch (const CollisionCheckerException & e) {
+    RCLCPP_ERROR(rclcpp::get_logger(name_), "%s", e.what());
+    return false;
+  }
+}
+
+double CollisionChecker::scorePose(
+  const geometry_msgs::msg::Pose2D & pose)
+{
+  try {
+    costmap_ = costmap_sub_->getCostmap();
+  } catch (const std::runtime_error & e) {
+    throw CollisionCheckerException(e.what());
+  }
+
+  unsigned int cell_x, cell_y;
+  if (!costmap_->worldToMap(pose.x, pose.y, cell_x, cell_y)) {
+    RCLCPP_DEBUG(rclcpp::get_logger(name_), "Map Cell: [%d, %d]", cell_x, cell_y);
+    throw IllegalPoseException(name_, "Pose Goes Off Grid.");
+  }
+
+  Footprint footprint;
+  if (!footprint_sub_->getFootprint(footprint)) {
+    throw CollisionCheckerException("Footprint not available.");
+  }
+
+  Footprint footprint_spec;
+  unorientFootprint(footprint, footprint_spec);
+  transformFootprint(pose.x, pose.y, pose.theta, footprint_spec, footprint);
+
+  // now we really have to lay down the footprint in the costmap_ grid
+  unsigned int x0, x1, y0, y1;
+  double line_cost = 0.0;
```

Comment:
`line_cost` could be declared inside the for loop

Reply:
Why would we want to do that here?